enterprise/backend/src/metabase_enterprise/serialization/v2/ingest/yaml.clj

```diff
@@ -1,44 +1,66 @@
+  "Note that throughout the YAML file handling, the `:serdes/meta` abstract path is referred to as the \"hierarchy\",
+  to avoid confusion with filesystem paths."
   (:require [clojure.java.io :as io]
             [metabase-enterprise.serialization.v2.ingest :as ingest]
-            [yaml.core :as yaml])
-  (:import java.io.File))
+            [metabase-enterprise.serialization.v2.utils.yaml :as u.yaml]
+            [metabase.util.date-2 :as u.date]
+            [yaml.core :as yaml]
+            [yaml.reader :as y.reader])
+  (:import java.io.File
+           java.time.temporal.Temporal))
 
-(defmulti ^:private build-metas
-  (fn [^File file] (.getName file)))
+(extend-type Temporal y.reader/YAMLReader
+  (decode [data]
+    (u.date/parse data)))
 
-(defmethod build-metas "settings.yaml" [file]
+(defn- build-settings [file]
   (let [settings (yaml/from-file file)]
     (for [[k _] settings]
-      {:model "Setting" :id (name k)})))
-
-(defmethod build-metas :default [^File file]
-  (let [model-name   (-> file .getParentFile .getName)
-        [_ id label] (re-matches #"^([A-Za-z0-9_-]+)(?:\+(.*))?.yaml$" (.getName file))]
-    [(cond-> {:model model-name :id id}
-       label (assoc :label label))]))
-
-(defn- ingest-entity [root-dir {:keys [model id label] :as meta-map}]
-  (let [filename (if label
-                   (str id "+" label ".yaml")
-                   (str id ".yaml"))]
-    (-> (io/file root-dir model filename)
-        yaml/from-file
-        (assoc :serdes/meta meta-map))))
+      ; We return a path of 1 item, the setting itself.
+      [{:model "Setting" :id (name k)}])))
+
+
+(defn- build-metas [^File root-dir ^File file]
+  (let [path-parts (u.yaml/path-split root-dir file)]
+    (if (= ["settings.yaml"] path-parts)
+      (build-settings file)
+      [(u.yaml/path->hierarchy path-parts)])))
+
+(defn- read-timestamps [entity]
+  (->> (keys entity)
+       (filter #(.endsWith (name %) "_at"))
+       (reduce #(update %1 %2 u.date/parse) entity)))
+
+(defn- ingest-entity [root-dir hierarchy]
+  (let [entity    (-> (u.yaml/hierarchy->file root-dir hierarchy)
+                      yaml/from-file
+                      (assoc :serdes/meta hierarchy)
+                      (read-timestamps))
+        ;; Strip the labels off the hierarchy; they might have been manipulated (eg. truncated) by the storage system.
+        hierarchy (mapv #(dissoc % :label) hierarchy)]
+    (assoc entity :serdes/meta hierarchy)))
 
 (deftype YamlIngestion [^File root-dir settings]
   ingest/Ingestable
   (ingest-list [_]
     (eduction (comp (filter (fn [^File f] (.isFile f)))
-                    (mapcat build-metas))
+                    (mapcat (partial build-metas root-dir)))
               (file-seq root-dir)))
-  (ingest-one [_ {:keys [model id] :as meta-map}]
-    (if (= "Setting" model)
-      {:serdes/meta meta-map :key (keyword id) :value (get settings (keyword id))}
-      (ingest-entity  root-dir meta-map))))
+
+  (ingest-one [_ meta-maps]
```

Comment:
```suggestion
  (ingest-one [_ hierarchies]
```
for consistency?

Reply:
I've mostly settled on `path` in the latest edition. I agree it's not the best name, but I can't find anything else I like better.